### ISSUE TITLE
Port twisted.cred.credentials to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6176]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6176|
|Type|enhancement|
|Created|2012-11-08 19:21:04Z|
|Branch|https://github.com/twisted/twisted/tree/tccredentials-py3-6176-3|

`twisted.cred.credentials` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6176 6176
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_tccredentials_py3_6176_3 branches/tccredentials-py3-6176-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1352402464000000 1352402464000000
changetime__1431308050953384 1431308050953384
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ihrachyshka set owner to ihrachyshka<br>ihrachyshka set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|ihrachyshka removed owner<br>ihrachyshka set status to `new`|
|-|-|

This also depends on [#5203](https://github.com/twisted/twisted/issues/5203) being fixed (unicode vs. str in py3k is not considered in filepath code).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [44283]) Branching to tccredentials-py3-6176.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl removed owner|
|-|-|

Buildbots are green, coverage looks good, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for working on this. Changes looks good.

Thistedchecker is red...  at least networkFormat needs a docstring

-------

I know that this is not specific to current changes.

I am a bit nervous while reviewing this as I see that some tests just reuse the logic from the implementation so for example let's say that binascii.hexlify is broken, than the test will still pass as it uses the same broken implementation. Most of the time this is ok, but not for code related to authentication.

So instead of computing expected in test code using the same methods and logic from the implementation I was expecting to see hardcoded expected values.

-------

just minor comments

0. There is no news file.

1. Hope that md5-sess fix will land soon and we can remove it from the diff

2. I see the logic for `__all3__` duplicated in many test module, maybe we need to refactor it into a helper method... not what all tests suites ends with 'Tests' maybe we can have something like `partialPy3Ported('SomeTests', 'AnotherTests') and the code can automatically remove all other tests suites. 

3. For `class Testable(components.Adapter):` maybe instead of pass we can add a docstring.

4. Maybe you can implement networkFormat as `networkFormat(format, *args)` so that we don't need to put the tuple each time.

-----

Thanks!

Please merge md5-sess fix and fix twistedchecker + newsfile and then resubmit for review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [44322]) Branching to tccredentials-py3-6176-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Thanks Adi!

- Fixed the TwistedChecker errors.
- Added a newsfile.
- Removed the md5-sess from the diff when I merged forward.
- Reimplemented networkFormat as a bytestring join operation, since it's much faster, and doesn't do encoding dances as much. 
- I think args vs *args is clearer... not sure.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

Changes looks good and I am +1 for merging them.

The newsfile is a bit vague to me. Maybe it should talk about the credential checkers available on Py3 as not all credential checkers are available yet.

-----

I am still not happy with code duplication from `__all3__` part and the manual work need to update the `__all__` list. I don't think that this should stop this merge.

What do you say if we have something like this in compat... implemented in a separate ticket?

```
# Defined in compat
def testSuitesPortedToPython3(py3Names):
    """
    Should be call at the end of a module.
    """
    allNames = []
    for key, value in globals().items():
        if value is unittest.TestCase:
            continue
        if isinstance(value, unittest.TestCase):
            allNames.append(key)

    if not allNames:
        raise AssertionError('No test suite. Did you call this at the end of the module?')

    for name in allNames:
        if name not in py3Names:
            del globals()[name]


# called at the end of test module
testSuitesPortedToPython3(["CramMD5CredentialsTestCase"])
```


since this is security related code I will leave this for another dev to review it.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

(In [44584]) Branching to tccredentials-py3-6176-3.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_13"></a>|@moshez commented|
|-|-|

In joinBytes, what is the use case for decoding ASCII rather than erroring out on unicode immediately?
I would rather if someone tries passing in unicode, that it break cleanly rather than waiting for the first French person to come along. Worst, I'm not entirely sure this doesn't introduce a new and exciting side-channel attack by timing the exception. I'm not sure that it does, but I'm scared.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

The use case for decoding ASCII is the same as ```twisted.python.compat.networkString``` -- sometimes things give you ASCII encoded in a text string (like, an IP address). Plus, twisted.cred (should) only ever get bytes -- eg. from twisted.web.

As far as the side-channel attack... I'm not sure that we're side-channel attack resistant, anywhere. Especially when you put PyPy into the mix. :P

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks for the work on this, hawkowl.

  1. I am also pretty down on `joinBytes`.  At every call site, there are bits of network input that are correctly typed as bytes (`nonce`) mixed in with other bits of input that are either clearly ASCII-only native strings (`k`, `now`, `clientip`) or API inputs which might easily incorrectly be non-ASCII unicode (`.username`, `.realm`, `.password`, `v`).  It seems to me that the only function of `joinBytes` is to obscure encoding errors and make it harder to track down which one is the one that wasn't encoded correctly, by putting the encoding into a loop instead of a nice traceback pointing at the problematic line which is decoding a specific variable.  I think this should be carefully done at each place where it's called, and I _definitely_ don't want to expose a public API to encourage other code to start doing this.
  2. Minor stylistic nitpick; I think it should be `from binascii import hexlify` rather than `binascii.hexlify`.
  3. `CramMD5Credentials` needs attribute documentation for `challenge` and `response`.  Sorry for the documentation burden, but this is definitely starting to be a subtle change.
  4. Please put a comment around the `except (ImportError, SyntaxError)` explaining what that's doing (presumably pamauth isn't ported)
  5. I don't understand the need for the `TypeError` in the `from twisted.cred import checkers` import; can we get rid of this?  Adding blanket exception handlers around imports like this can make debugging really annoying.  Presumably it's importable after the changes in this branch...

Some optional notes:

  1. I think we should generally prefer `.format` rather than `%` when we're formatting strings.
  2. Since you're changing types to `bytes`, might as well change it to `L{bytes}` rather than `C{bytes}` everywhere.
  3. I like adi's idea for a utility decorator making the manual (temporary) `__all__` munging more declarative.  Can you file a ticket for this?

I think this needs a re-review to see how it looks with `joinBytes` removed, but otherwise I think it is close to ready.  I have really mixed feelings about `username` and `password` being bytes, but ... these data structures probably need to be re-invented from scratch to accommodate text input :-\.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl removed owner|
|-|-|

Okay, I think I've addressed nearly everything.

I think joinBytes is a bit misguided, yeah, and I've replaced it with b"whatever".join(). In the tests I've tried to replace them with ```thing + b":" + spam``` style to make it less likely that a formatting bug would slip through (as they were being constructed two different ways and must be the same).

I'll file a ticket for the ```__all__``` stuff.

Builders are spun and a wonderful green, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks!

You did miss the `binascii.hexlify` → `hexlify` thing, but I guess that's fine.

In addition to filing a ticket for making the `__all__` thing declarative, I realize that this ugliness is part of the same problem on the incremental tickets I was reviewing yesterday.  It's simply too easy to miss this conditional and to assume that we have more coverage than we do, because the test module is marked as "ported".

I also think that the `TypeError` on import thing is mostly an accident and even with the explanation I don't like it; test code should have as few control structures in it as possible, and this adds a whole dimension for partial failure that isn't really valid.

But, okay, actionable feedback:

  1. I think that we should take the opportunity here to create a `twisted.cred.test` package and separate out the different module tests.  So we should move `CramMD5CredentialsTests` and to a new `twisted.cred.test.test_cramauth` module, and perhaps take the opportunity as well (since this _won't_ produce diff churn) to move `test_digestauth` to this package as well.  Then `test_cramauth` and `test_digestauth`  can be in dist3 and `test_newcred` won't be.  This means you'll need to make `CramMD5CredentialsTests` twistedchecker-clean and write some docstrings, but that's probably _less_ overall churn than the other unrelated changes in `test_newcred` to give it the illusion that it can be imported.
  2. Given the fresh perspective that seeing which test modules will actually be included, it is clear we are missing direct test coverage of our more basic credential types, like `UsernamePassword`.  Maybe add a new test module, `test_simpleauth`, just to make sure the constructors and `checkPassword` and various provider/implementor checks and such work as expected.

One more submit for re-review.  Thanks for slogging through this one; I feel like we are converging on a slightly better way to get these modules cleaned up and ported now.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl removed owner|
|-|-|

Hi Glyph, thanks for the review!

- I actually missed those ones, fixed that.
- I moved the CRAM tests into their own file, and moved all the cred tests (test_digestauth and test_newcred) into the twisted.cred.test package.
- I removed all of the churn from test_newcred.
- Added tests for the basic things that I thought I could test well enough. The SSH stuff is excluded, since I am not sure about what it does or what it should do.
- The PAM stuff is untested and I found a ticket advocating for its deprecation for horrendous security reasons. As such, I'm leaving it untouched.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @hawkowl|
|-|-|

Hi hawkie. Thanks. Ready to merge but for a few minor issues:

  1. There are numerous [twistedchecker errors](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2928/steps/run-twistedchecker/logs/new%20twistedchecker%20errors), but only one (in `twisted.cred.test`) is in added rather than moved code.
    1. Please fix the one new error.
    2. Please file a fast-following ticket to fix the remainder of the warnings in `test_digestauth` since this stuff is fresh in your mind. `test_newcred` is probably too much to ask for though :-).
  2. Since the file is moving anyway, could you please rename `test_newcred` to `test_cred` in its new home?  The "new" in the name conveys absolutely zero information.

Thanks, please land when done.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_20"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44731]) Merge tccredentials-py3-6176-3: Port twisted.cred.credentials to Python 3

Author: hawkowl
Reviewers: adiroiban, glyph
Fixes: [#6176](https://github.com/twisted/twisted/issues/6176)
